larpandoracontent/LArTwoDReco/TrackInEMShowerAlgorithm.cc

```diff
@@ -516,224 +510,188 @@ const Cluster* TrackInEMShowerAlgorithm::RefineTrack(const Cluster *const pClust
-    TwoDSlidingFitResultMap &microSlidingFitResultMap, TwoDSlidingFitResultMap &macroSlidingFitResultMap,  ClusterVector &clusterVector) const
+bool TrackInEMShowerAlgorithm::IsClusterRemnantDisconnected(const Cluster *const pRemnantCluster) const
 {
-    // Merge together cosmic ray muon pair
-    const Cluster *const pClusterToEnlarge(clusterAssociation.GetUpstreamCluster());
-    const Cluster *const pClusterToDelete(clusterAssociation.GetDownstreamCluster());
-
-    this->UpdateForClusterDeletion(pClusterToDelete, clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);    
-    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::MergeAndDeleteClusters(*this, pClusterToEnlarge, pClusterToDelete));
+    if (pRemnantCluster->GetNCaloHits() == 1)
+        return false;
 
-    // Merge in all collected hits from shower region between the pair
-    ClusterVector clustersToFragment;
-    for (const auto &entry : clusterToCaloHitListMap)
-        clustersToFragment.push_back(entry.first);
+    unsigned int count(0);
+    float maxSeparationDistance(-std::numeric_limits<float>::max());
 
-    // In the case of a delta ray split
-    if (clustersToFragment.empty())
+    const OrderedCaloHitList &orderedCaloHitList(pRemnantCluster->GetOrderedCaloHitList());
+    CartesianVector previousHitPosition((*orderedCaloHitList.begin()->second->begin())->GetPositionVector());
+    for (const OrderedCaloHitList::value_type &mapEntry : orderedCaloHitList)
     {
-        this->UpdateAfterClusterModification(pClusterToEnlarge, clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
-        return;
-    }
+        for (const CaloHit *const pCaloHit : *mapEntry.second) 
+        {
+            ++count;
+            
+            if (count == 1)
+                continue;
+            
+            const float separationDistance(std::sqrt(pCaloHit->GetPositionVector().GetDistanceSquared(previousHitPosition)));
 
-    std::sort(clustersToFragment.begin(), clustersToFragment.end(), LArClusterHelper::SortByNHits);
-    for (const Cluster *const pCluster : clustersToFragment)
-    {
-        this->UpdateForClusterDeletion(pCluster, clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
-        
-        const CaloHitList &caloHitsToMerge(clusterToCaloHitListMap.at(pCluster));
-        this->MergeCluster(pCluster, pClusterToEnlarge, caloHitsToMerge, clusterAssociation);
+            if (separationDistance > maxSeparationDistance)
+                maxSeparationDistance = separationDistance;
+
+            previousHitPosition = pCaloHit->GetPositionVector();
+        }
     }
-        
-    this->UpdateAfterClusterModification(pClusterToEnlarge, clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
+
+    if (maxSeparationDistance > m_maxHitSeparationForConnectedCluster)
```

Comment:
Could presumably bug out of the loop above, and stop making comparisons (i.e. return true from within loop), if any separation distance crosses this threshold?

Reply:
thanks!